File: scripts/core/i18n.js
/**
 * 汉化处理模块
 * 读取 opencode-i18n 配置并应用到源码
 */

const fs = require('fs');
const path = require('path');
const { glob } = require('glob');
const { getI18nDir, getOpencodeDir } = require('./utils.js');
const { step, success, error, indent } = require('./colors.js');

class I18n {
  constructor() {
    this.i18nDir = getI18nDir();
    this.opencodeDir = getOpencodeDir();
  }

  /**
   * 读取所有汉化配置文件
   */
  loadConfig() {
    if (!fs.existsSync(this.i18nDir)) {
      throw new Error(`汉化配置目录不存在: ${this.i18nDir}`);
    }

    const configs = [];

    // 遍历 opencode-i18n 目录
    const entries = fs.readdirSync(this.i18nDir, { withFileTypes: true });

    for (const entry of entries) {
      if (entry.isDirectory()) {
        const categoryDir = path.join(this.i18nDir, entry.name);
        const jsonFiles = glob.sync('*.json', { cwd: categoryDir });

        for (const file of jsonFiles) {
          const filePath = path.join(categoryDir, file);
          try {
            const content = JSON.parse(fs.readFileSync(filePath, 'utf-8'));
            configs.push({
              category: entry.name,
              fileName: file,
              ...content
            });
          } catch (err) {
            console.warn(`警告: 跳过无效配置 ${filePath}: ${err.message}`);
          }
        }
      }
    }

    return configs;
  }

  /**
   * 应用单个配置文件的替换规则
   */
  applyConfig(config) {
    // 使用 'file' 字段（不是 'targetFile'）
    if (!config.file || !config.replacements) {
      return { files: 0, replacements: 0 };
    }

    // OpenCode 源码在 packages/opencode/ 目录
    // 如果路径不是以 packages/ 开头，自动添加前缀
    let relativePath = config.file;
    if (!relativePath.startsWith('packages/')) {
      relativePath = path.join('packages/opencode', relativePath);
    }

    const targetPath = path.join(this.opencodeDir, relativePath);

    if (!fs.existsSync(targetPath)) {
      // 静默跳过不存在的文件
      return { files: 0, replacements: 0 };
    }

    let content = fs.readFileSync(targetPath, 'utf-8');
    // 规范化换行符：统一使用 LF
    content = content.replace(/\r\n/g, '\n');
    let replaceCount = 0;
    const originalContent = content;

    // replacements 是键值对对象
    for (const [find, replace] of Object.entries(config.replacements)) {
      // 也规范化查找字符串中的换行符
      const normalizedFind = find.replace(/\r\n/g, '\n');

      // 判断是否为简单单词（只包含字母和数字）
      const isSimpleWord = /^[a-zA-Z0-9]+$/.test(normalizedFind);

      if (isSimpleWord) {
        // 简单单词使用单词边界，避免误翻译代码标识符
        // 例如: "Status" 不会匹配 "DialogStatus" 中的 "Status"
        const wordBoundaryPattern = new RegExp(`\\b${normalizedFind}\\b`, 'g');
        if (wordBoundaryPattern.test(content)) {
          content = content.replace(wordBoundaryPattern, replace);
          replaceCount++;
        }
      } else {
        // 复杂模式（含特殊字符）使用普通替换
        if (content.includes(normalizedFind)) {
          content = content.replaceAll(normalizedFind, replace);
          replaceCount++;
        }
      }
    }

    if (content !== originalContent) {
      fs.writeFileSync(targetPath, content, 'utf-8');
      console.log(`  ✓ ${config.file} (${replaceCount} 处替换)`);
    }

    return { files: 1, replacements: replaceCount };
  }

  /**
   * 应用所有汉化配置
   */
  async apply(options = {}) {
    const { silent = false } = options;

    if (!silent) {
      step('应用汉化配置');
    }

    const configs = this.loadConfig();

    if (configs.length === 0) {
      throw new Error('未找到任何汉化配置文件');
    }

    if (!silent) {
      console.log(`找到 ${configs.length} 个配置文件`);
    }

    let totalFiles = 0;
    let totalReplacements = 0;

    for (const config of configs) {
      const result = this.applyConfig(config);
      totalFiles += result.files;
      totalReplacements += result.replacements;
    }

    if (!silent) {
      success(`汉化应用完成: ${totalFiles} 个文件, ${totalReplacements} 处替换`);
    }

    return { files: totalFiles, replacements: totalReplacements };
  }

  /**
   * 验证配置完整性
   */
  validate() {
    const configs = this.loadConfig();
    const errors = [];

    for (const config of configs) {
      if (!config.file) {
        errors.push(`${config.category}/${config.fileName}: 缺少 file 字段`);
      }
      if (!config.replacements || Object.keys(config.replacements).length === 0) {
        errors.push(`${config.category}/${config.fileName}: 缺少 replacements`);
      }
    }

    return errors;
  }

  /**
   * 获取汉化统计信息
   */
  getStats() {
    const configs = this.loadConfig();
    const stats = {
      totalConfigs: configs.length,
      categories: {},
      totalReplacements: 0,
    };

    for (const config of configs) {
      const category = config.category;
      if (!stats.categories[category]) {
        stats.categories[category] = { count: 0, replacements: 0 };
      }
      stats.categories[category].count++;
      if (config.replacements) {
        const count = Object.keys(config.replacements).length;
        stats.categories[category].replacements += count;
        stats.totalReplacements += count;
      }
    }

    return stats;
  }
}

module.exports = I18n;

File: cli-go/internal/core/i18n.go
package core

import (
	"embed"
	"encoding/json"
	"fmt"
	"io/fs"
	"os"
	"path/filepath"
	"regexp"
	"strings"
)

//go:embed assets/opencode-i18n
var embeddedAssets embed.FS

// TranslationConfig 汉化配置结构
type TranslationConfig struct {
	Category     string
	FileName     string
	ConfigPath   string
	File         string            `json:"file"`
	Replacements map[string]string `json:"replacements"`
}

// Replacement 单条替换规则（用于 verify 命令）
type Replacement struct {
	From string
	To   string
}

// GetReplacementsList 获取替换规则列表
func (c *TranslationConfig) GetReplacementsList() []Replacement {
	var list []Replacement
	for from, to := range c.Replacements {
		list = append(list, Replacement{From: from, To: to})
	}
	return list
}

// LoadI18nConfig 从单个 JSON 文件加载配置
func LoadI18nConfig(path string) (*TranslationConfig, error) {
	var config TranslationConfig
	if err := ReadJSON(path, &config); err != nil {
		return nil, err
	}
	config.ConfigPath = path
	return &config, nil
}

// I18n 汉化处理器
type I18n struct {
	i18nDir     string
	opencodeDir string
	useEmbedded bool
}

// NewI18n 创建 I18n 实例
func NewI18n() (*I18n, error) {
	i18nDir, err := GetI18nDir()
	useEmbedded := false

	// 如果获取目录失败或目录不存在，尝试使用内置资源
	if err != nil || !DirExists(i18nDir) {
		useEmbedded = true
		i18nDir = "assets/opencode-i18n" // embedded 中的相对路径
	}

	opencodeDir, err := GetOpencodeDir()
	if err != nil {
		// 如果连 OpenCode 源码目录都找不到，那就真的无法继续了
		return nil, err
	}

	if useEmbedded {
		fmt.Println("提示: 使用内置汉化配置")
	} else {
		fmt.Printf("提示: 使用外部汉化配置: %s\n", i18nDir)
	}

	return &I18n{
		i18nDir:     i18nDir,
		opencodeDir: opencodeDir,
		useEmbedded: useEmbedded,
	}, nil
}

// LoadConfig 读取所有汉化配置文件
func (i *I18n) LoadConfig() ([]TranslationConfig, error) {
	var configs []TranslationConfig
	var entries []fs.DirEntry
	var err error

	if i.useEmbedded {
		entries, err = fs.ReadDir(embeddedAssets, i.i18nDir)
	} else {
		entries, err = os.ReadDir(i.i18nDir)
	}

	if err != nil {
		return nil, err
	}

	for _, entry := range entries {
		if entry.IsDir() {
			categoryName := entry.Name()

			var files []fs.DirEntry
			if i.useEmbedded {
				// Embedded FS 路径必须使用正斜杠
				embedPath := i.i18nDir + "/" + categoryName
				files, err = fs.ReadDir(embeddedAssets, embedPath)
			} else {
				categoryDir := filepath.Join(i.i18nDir, categoryName)
				files, err = os.ReadDir(categoryDir)
			}

			if err != nil {
				continue
			}

			for _, file := range files {
				if strings.HasSuffix(file.Name(), ".json") {
					var config TranslationConfig
					var configPath string
					var readErr error

					if i.useEmbedded {
						configPath = i.i18nDir + "/" + categoryName + "/" + file.Name()
						var data []byte
						data, readErr = fs.ReadFile(embeddedAssets, configPath)
						if readErr == nil {
							readErr = json.Unmarshal(data, &config)
						}
					} else {
						configPath = filepath.Join(i.i18nDir, categoryName, file.Name())
						readErr = ReadJSON(configPath, &config)
					}

					if readErr != nil {
						fmt.Printf("警告: 解析配置文件失败 %s: %v\n", configPath, readErr)
						continue
					}
					config.Category = categoryName
					config.FileName = file.Name()
					config.ConfigPath = configPath
					configs = append(configs, config)
				}
			}
		}
	}

	return configs, nil
}

// ApplyResult 应用结果
type ApplyResult struct {
	File         string
	Success      bool
	Replacements struct {
		Total   int
		Success int
		Failed  int
	}
	Skipped    bool
	SkipReason string
}

// ApplyConfig 应用单个配置文件的替换规则
func (i *I18n) ApplyConfig(config TranslationConfig, dryRun bool) ApplyResult {
	result := ApplyResult{
		File: config.File,
	}

	if config.File == "" || len(config.Replacements) == 0 {
		result.Skipped = true
		result.SkipReason = "缺少 file 或 replacements 字段"
		return result
	}

	relativePath := config.File
	if !strings.HasPrefix(relativePath, "packages/") {
		relativePath = filepath.Join("packages", "opencode", relativePath)
	}

	targetPath := filepath.Join(i.opencodeDir, relativePath)

	if !Exists(targetPath) {
		result.Skipped = true
		result.SkipReason = "目标文件不存在"
		return result
	}

	contentBytes, err := os.ReadFile(targetPath)
	if err != nil {
		result.Skipped = true
		result.SkipReason = fmt.Sprintf("读取文件失败: %v", err)
		return result
	}
	content := string(contentBytes)
	// 规范化换行符
	content = strings.ReplaceAll(content, "\r\n", "\n")
	originalContent := content

	result.Replacements.Total = len(config.Replacements)

	for find, replace := range config.Replacements {
		// 规范化查找字符串
		normalizedFind := strings.ReplaceAll(find, "\r\n", "\n")

		// 判断是否为简单单词（只包含字母和数字）
		isSimpleWord, _ := regexp.MatchString("^[a-zA-Z0-9]+$", normalizedFind)

		matched := false
		if isSimpleWord {
			// 简单单词使用单词边界
			wordBoundaryPattern := regexp.MustCompile(`\b` + regexp.QuoteMeta(normalizedFind) + `\b`)
			if wordBoundaryPattern.MatchString(content) {
				if !dryRun {
					content = wordBoundaryPattern.ReplaceAllString(content, replace)
				}
				matched = true
			}
		} else {
			// 复杂模式使用普通替换
			if strings.Contains(content, normalizedFind) {
				if !dryRun {
					content = strings.ReplaceAll(content, normalizedFind, replace)
				}
				matched = true
			}
		}

		if matched {
			result.Replacements.Success++
		} else {
			result.Replacements.Failed++
		}
	}

	if !dryRun && content != originalContent {
		if err := os.WriteFile(targetPath, []byte(content), 0644); err != nil {
			result.Success = false
			fmt.Printf("错误: 写入文件失败 %s: %v\n", targetPath, err)
		} else {
			result.Success = result.Replacements.Success > 0
		}
	} else {
		result.Success = result.Replacements.Success > 0
	}

	return result
}
